targets/TARGET_Silicon_Labs/TARGET_EFM32/trng/sl_trng.c

```diff
@@ -0,0 +1,373 @@
+        if ( status & TRNG_STATUS_REPCOUNTIF )
+        {
+            return SL_TRNG_ERR_REPETITION_COUNT_TEST_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT64IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_64_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT4096IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_4096_FAILED;
+        }
+        if ( status & TRNG_STATUS_ALMIF )
+        {
+            return SL_TRNG_ERR_NOISE_ALARM;
+        }
+    }
+
+    return 0;
+}
+
+int sl_trng_check_entropy( TRNG_TypeDef *device )
+{
+    volatile uint32_t val32;
+    int i, ret = 0;
+    uint32_t ctrl = device->CONTROL;
+
+    /* Setup control register */
+    device->CONTROL =
+        TRNG_CONTROL_ENABLE |
+        TRNG_CONTROL_REPCOUNTIEN |
+        TRNG_CONTROL_APT64IEN |
+        TRNG_CONTROL_APT4096IEN |
+        TRNG_CONTROL_PREIEN |
+        TRNG_CONTROL_ALMIEN;
+
+    /* Apply software reset */
+    sl_trng_soft_reset(device);
+
+    /* Check FIFO level is non-zero . */
+    for (i=0; i<FIFO_LEVEL_RETRY; i++)
+    {
+        if ( device->FIFOLEVEL )
+        {
+            break;
+        }
+    }
+    /* Check for no data within timeout (max retry count) */
+    if (i>=FIFO_LEVEL_RETRY)
+    {
+        ret = SL_TRNG_ERR_NO_DATA;
+    }
+    else
+    {
+        /* Read at least 4097x2 bits (~257 x 32 bits) in order for the longest
+           test to complete (adaptive proportion test of 4096 samples). */
+        for (i=0; i<TEST_WORDS_MIN; i++)
+        {
+            val32 = device->FIFO;
+            (void)val32;
+        }
+
+        /* Check in status register for errors. */
+        ret = sl_trng_check_status( device );
+    }
+
+    /* Restore initial value of control register */
+    device->CONTROL = ctrl;
+
+    return ret;
+}
+
+static void sl_trng_read_chunk( TRNG_TypeDef *device,
+                                unsigned char *output,
+                                size_t len )
+{
+    uint32_t * out32 = (uint32_t *) output;
+    uint32_t tmp;
+
+    /* Read known good available data. */
+    while ( len >= 4)
+    {
+        *out32++ = device->FIFO;
+        len -= 4;
+    }
+
+    /* Handle the case where len is not a multiple of 4. */
+    if ( len < 4 )
+    {
+        tmp = device->FIFO;
+        memcpy((uint8_t *)out32, (const uint8_t *) &tmp, len);
+    }
+}
+
+int sl_trng_poll( TRNG_TypeDef *device,
+                  unsigned char *output,
+                  size_t len,
+                  size_t *olen )
+{
+    size_t output_len = 0;
+    size_t count = 0;
+    size_t available;
+    int ret = 0;
+
+    while (len > 0)
+    {
+        available = device->FIFOLEVEL * 4;
+        if (available == 0)
+        {
+            break;
+        }
+
+#if !defined(SL_TRNG_IGNORE_ALL_ALARMS)
+        /* Check status for current data in FIFO
+         * and handle any error conditions. */
+        ret = sl_trng_check_status( device );
```

Comment:
Well, on a noise alarm you really should throw away the generated data and reset the TRNG (which is already happening in check_status). Thus, the polling 'failed'. IMO, this should bubble up to the user, who can then decide whether or not to retry.
The other statuses are irrecoverable in the sense that they point to a larger issue, most likely someone tampering with the TRNG ring oscillators.